File: src/endpoints/file/get_index.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.libs.object import json_parse
from src.modules.verify import checking_token

async def get_index(params: dict = {}, req = None, _db: DB = None, _token = None):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        token = checking_token(req, db, use_public=True) if not _token else _token

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # ser where
        where = []
        if params.module:
            where.append(f'and module LIKE \'{params.module}\'')
        if params.module and params.module_srl:
            where.append(f'and module_srl = {params.module_srl}')
        if params.name:
            where.append(f'and name LIKE "%{params.name}%"')
        if params.mime:
            where.append(f'and mime LIKE "%{params.mime}%"')

        # get total
        total = db.get_count(
            table_name = Table.FILE.value,
            where = where,
        )
        if not (total > 0): raise Exception('No data', 204)

        # get index
        index = db.get_items(
            table_name=Table.FILE.value,
            fields=fields,
            where=where,
            limit={ 'size': params.size, 'page': params.page },
            order={ 'order': params.order, 'sort': params.sort },
            unlimited=params.unlimited,
        )
        def transform_item(item: dict) -> dict:
            if 'json' in item:
                item['json'] = json_parse(item['json'])
            if 'path' in item:
                del item['path']
            return item
        index = [ transform_item(item) for item in index ]

        # set result
        result = output.success({
            'message': 'Complete get file index.',
            'data': {
                'total': total,
                'index': index,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/file/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import Service from '@/classes/Service'
import { parseJSON } from '@/libs/objects'
import type { FileModel } from './__model'

type GetIndexParams = {
  query: FileModel['getIndexQuery']
  service: Service
}

export default async function getIndex({ query, service }: GetIndexParams)
{
  try
  {
    // set assets
    const _table = `${DB.TABLE.FILE} AS f`
    const _field = query.field ? query.field.split(',') : ''
    let _where: string[] = []
    let _values: ZZ = {}
    let _join: string[] = []
    if (query.module)
    {
      _where.push(`AND module LIKE $module`)
      _values['$module'] = query.module
    }
    if (query.module && query.module_srl)
    {
      _where.push(`AND module_srl = $module_srl`)
      _values['$module_srl'] = query.module_srl
    }
    if (query.name)
    {
      _where.push(`AND name LIKE '%' || $name || '%'`)
      _values['$name'] = query.name
    }
    if (query.mime)
    {
      _where.push(`AND mime LIKE '%' || $mime || '%'`)
      _values['$mime'] = query.mime
    }

    // get count
    const count = db.getCount({
      table: _table,
      where: _where,
      join: _join,
      values: _values,
    })
    if (count.data <= 0) throw new ServiceError('No data', { status: 204 })

    // get index
    const index = db.getIndex({
      table: _table,
      field: _field,
      where: _where,
      join: _join,
      order: Boolean(query.order || query.sort) ? query.order : 'srl',
      sort: Boolean(query.order || query.sort) ? query.sort : 'desc',
      page: query.page ?? 1,
      size: query.size ?? service.preference['file.index.size'],
      values: _values,
    })
    const _index = index.data.map((o: ZZ) => {
      return {
        ...o,
        json: o.json ? parseJSON(o.json) : undefined,
        path: undefined,
      }
    })

    return {
      total: count.data,
      index: _index,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get File index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
